### PR TITLE
node json parse error fix

### PR DESCRIPTION
**Pull request by alleycat-at-git**
The communication between rails server and node server sometimes breaks. This PR fixes this issue.

Improvements to be made:
1) The process now uses eof symbol to signal the end of input for a request / response. While it's very unlikely to break, it is not 100% safe method. The safe implementation is to transmit the number of bytes in request / response
2) Node server doesn't restart when the server bundle changes.
3) You have to remove manually node.sock after node server stops.
4) Render now uses eval, which means that all async calls inside the server bundle becomes sync calls

---

This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/532)

### CODE REVIEW

**Comment by alleycat-at-git** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[lib/generators/react_on_rails/templates/node/base/client/node/server.js, line 14 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/532#-KPn-XfBfrDOUZApFst9:-KPn-XfBfrDOUZApFstA:b-3sotxw) ([raw file](https://github.com/shakacode/react_on_rails/blob/c1eeccb25846ef95477ee7bd11c37601ff39e221/lib/generators/react_on_rails/templates/node/base/client/node/server.js#L14)):*
> ```JavaScript
> }
> 
> Handler.prototype.handle = function (connection) {
> ```

Note that fat arrow usage introduces a bug here - it binds this to be a global one, but this is required to be Handler inside this function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/532)*

File: lib/generators/react_on_rails/templates/node/base/client/node/server.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ const net = require('net');
 const fs = require('fs');
 
 const bundlePath = '../../app/assets/webpack/';
-let bundleFileName = 'webpack-bundle.js';
+let bundleFileName = 'server-bundle.js';
```
Comment (justin808) [inline lib/generators/react_on_rails/templates/node/base/client/node/server.js:5]: @AlexKVal This is supposed to work with the default generator. I don't think you should change the file name.

CC: @jbhatab

File: lib/generators/react_on_rails/templates/node/base/client/node/server.js [hunk 2]
```diff
@@ -30,7 +36,7 @@ Handler.prototype.handle = (connection) => {
   }
 };
 
-Handler.prototype.initialize = () => {
+Handler.prototype.initialize = function () {
```
Comment (justin808) [inline lib/generators/react_on_rails/templates/node/base/client/node/server.js:40]: why are you switching this? You'll get lint errors.

**Comment by alleycat-at-git** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[lib/generators/react_on_rails/templates/node/base/client/node/server.js, line 39 \[r1\]](https://reviewable.io:443/reviews/shakacode/react_on_rails/532#-KPr4VAPGkZ1_ZENwNf8:-KPr4VAPGkZ1_ZENwNf9:b-k0xcv8) ([raw file](https://github.com/shakacode/react_on_rails/blob/c1eeccb25846ef95477ee7bd11c37601ff39e221/lib/generators/react_on_rails/templates/node/base/client/node/server.js#L39)):*
You mean switch fat arrow for function? I mentioned above that fat arrow introduces a bug here. 

> Note that fat arrow usage introduces a bug here - it binds this to be a global one, but this is required to be Handler inside this function.

Also I don't see why `function` usage would produce lint errors. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react_on_rails/532)*

File: lib/generators/react_on_rails/templates/node/base/client/node/server.js (hunk 1 above)
Comment (AlexKVal) [reply lib/generators/react_on_rails/templates/node/base/client/node/server.js:5]: @justin808 it seems you meant @alleycat-at-git 😉

**Comment by justin808** [comment]
@alleycat-at-git Please fix these small lint issues and push.

```
npm run eslint -- --fix

rubocop -a
```


```
> eslint .
The react/require-extension rule is deprecated. Please use the import/extensions rule from eslint-plugin-import instead.
/home/travis/build/shakacode/react_on_rails/lib/generators/react_on_rails/templates/node/base/client/node/server.js
  14:28  warning  Missing function expression name  func-names
  15:20  warning  Missing function expression name  func-names
  40:32  warning  Missing function expression name  func-names
✖ 3 problems (0 errors, 3 warnings)
cd /home/travis/build/shakacode/react_on_rails && npm run jscs
> react-on-rails@6.1.0 jscs /home/travis/build/shakacode/react_on_rails
> jscs -e -v .
cd /home/travis/build/shakacode/react_on_rails && npm run flow
> react-on-rails@6.1.0 flow /home/travis/build/shakacode/react_on_rails
> flow check node_package
Found 0 errors
cd /home/travis/build/shakacode/react_on_rails && bundle exec rubocop .
Inspecting 93 files
........................................................................C....................
Offenses:
lib/react_on_rails/server_rendering_pool/node.rb:54:21: C: Style/SpaceAroundOperators: Surrounding space missing for operator =.
          eof_symbol="\r\n\0"
                    ^
lib/react_on_rails/server_rendering_pool/node.rb:57:36: C: Style/SpaceAroundOperators: Surrounding space missing for operator +.
            js_context.send(js_code+eof_symbol, 0)
                                   ^
lib/react_on_rails/server_rendering_pool/node.rb:59:19: C: Style/ParenthesesAroundCondition: Don't use parentheses around the condition of a while.
            while (result[-eof_symbol.length..-1] != eof_symbol)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
93 files inspected, 3 offenses detected
rake aborted!
Command failed with status (1): [cd /home/travis/build/shakacode/react_on_r...]
/home/travis/build/shakacode/react_on_rails/rakelib/task_helpers.rb:20:in `block in sh_in_dir'
/home/travis/build/shakacode/react_on_rails/rakelib/task_helpers.rb:20:in `each'
/home/travis/build/shakacode/react_on_rails/rakelib/task_helpers.rb:20:in `sh_in_dir'
/home/travis/build/shakacode/react_on_rails/rakelib/lint.rake:7:in `block (2 levels) in <top (required)>'
/home/travis/.rvm/gems/ruby-2.3.1/gems/rake-11.2.2/exe/rake:27:in `<top (required)>'
/home/travis/.rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `eval'
/home/travis/.rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => default => lint => lint:lint => lint:rubocop
(See full trace by running task with --trace)
The command "rake" exited with 1.
cache.2
store build cache
$ bundle clean
Cleaning all the gems on your system is dangerous! If you're sure you want to remove every system gem not in this bundle, run `bundle clean --force`.
0.00s
10.96schange detected (content changed, file is created, or file is deleted):
/home/travis/build/shakacode/react_on_rails/node_modules/archive-type/index.js
/home/travis/build/shakacode/react_on_rails/node_modules/archive-type/license
/home/travis/build/shakacode/react_on_rails/node_modules/archive-type/package.json
/home/travis/build/shakacode/react_on_rails/node_modules/archive-type/readme.md
/home/travis/build/shakacode/react_on_rails/node_modules/array-differ/index.js
/home/travis/build/shakacode/react_on_rails/node_modules/array-differ/package.json
/home/travis/build/shakacode/react_on_rails/node_modules/array-differ/readme.md
/home/travis/build/shakacode/react_on_rails/node_modules/babel-cli/node_modules/babylon/CHANGELOG.md
/home/travis/build/shakacode/react_on_rails/node_modules/babel-cli/node_modules/babylon/lib/parser/expression.js
/home/travis/build/shakacode/react_on_rails/node_modules/babel-cli/node_modules/babylon/lib/parser/lval.js
/home/travis/build/shakacode/react_on_rails/node_modules/babel-cli/node_modules/babylon/lib/parser/statement.js
/home/
...
changes detected, packing new archive
.
uploading archive
Done. Your build exited with 1.
```

**Comment by justin808** [comment]
@alleycat-at-git a few small lint issues. We should merge this.

**Comment by justin808** [comment]
* @alleycat-at-git we should get this merged
* @rocLv, please take a look
* We should consider using an express server rather than unix threads, to do the rendering, as is done with this solution: https://github.com/airbnb/hypernova. We should investigate how we could integrate or leverage that github repo. @rocLv This is the thing to research!

**Comment by justin808** [comment]
@alleycat-at-git can we release this as-is? I'd like to do a release this weekend.

**Comment by justin808** [comment]
CI is still broken. Let's get this merged.

**Comment by justin808** [comment]
Please update the CHANGELOG.md and fix the lint issues.

**Comment by alleycat-at-git** [comment]
see #641

**Pull request closed without merging**